File: 322. Coin Change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        sort(coins.begin(), coins.end(), greater<int>());
        int ret = INT_MAX;
        change(coins, amount, ret, 0, 0);
        return ret == INT_MAX? -1: ret;
    }

    void change(vector<int>& coins, int amount, int& ret, int idx, int ttl) {
        if (amount == 0) {
            ret = min(ret, ttl);
            return;
        }
        if (amount < 0 || idx >= coins.size()) return;
        change(coins, amount - coins[idx], ret, idx, ttl + 1);
        change(coins, amount, ret, idx + 1, ttl);
    }
};

File: 322. Coin Change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        if amount == 0: 
            return 0
        
        MAX = float(math.inf)
        dp = [MAX] * (amount + 1) # the minimum coin change used amount $i
        dp[0] = 0
        if 1 in coins:
            dp[1] = 1
        for i in range(2, amount + 1):
            for coin in coins:
                if i - coin >= 0 and dp[i-coin] != -1:
                    dp[i] = min(dp[i], dp[i-coin] + 1)
        if dp[amount] == MAX:
            return -1
        return dp[amount]
